document how to use alternate backends with our asymmetric abcs (#12881)
pyca/cryptography: docs/hazmat/primitives/asymmetric/cloudhsm.rst

.. hazmat::

> Cloud KMS and HSM asymmetric backends

Reviewer: ```suggestion
Cloud KMS and HSM Asymmetric Keys
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. testsetup::

    """
    We need to have this exist so the doctest below allows us to
    test that we're satisfying the base class requirements.
    """
    class Response:
        def __init__(self, signature):
            self.signature = signature

    class SomeCloudClient:
        def __init__(self, creds):
            pass

        def sign(self, key_id, algorithm, message):
            return Response(b"\x00" * (self.key_size(key_id) // 8))

        def key_size(self, key_id):
            return 2048

``cryptography`` provides a set of abstract base classes for asymmetric keys

> that can be used to integrate with cloud key management services, HSMs, et cetera.

Reviewer: ```suggestion
that can be used to integrate with cloud key management services, HSMs, and other ways of managing keys that are not in-memory.
```

if you only need a subset of functionality.

.. doctest::

    >>> import typing
    >>> from cryptography.hazmat.primitives.asymmetric import rsa, utils
    >>> from cryptography.hazmat.primitives import hashes, serialization
    >>> from cryptography.hazmat.primitives.asymmetric.padding import AsymmetricPadding, PKCS1v15
    >>>
    >>> class CloudRSAPrivateKey(rsa.RSAPrivateKey):
    ...     def __init__(self, creds, key_id):
    ...         self._creds = creds
    ...         self._cloud_client = SomeCloudClient(creds)
    ...         self._key_id = key_id
    ...
    ...     def sign(
    ...         self,
    ...         data: bytes,
    ...         padding: AsymmetricPadding,
    ...         algorithm: typing.Union[utils.Prehashed, hashes.HashAlgorithm],
    ...     ) -> bytes:
    ...         """
    ...         Signs data using the cloud KMS. You'll need to define a mapping
    ...         between the way your cloud provider represents padding and algorithms
    ...         and the way cryptography represents them.
    ...         """
    ...         # Map cryptography padding/algorithm to KMS signing algorithm

>     ...         kms_algorithm = self._map_to_kms_algorithm(padding, algorithm)

Reviewer: This is wrong for prehashed, there you need to unpack the inner algorithm.

    ...         else:
    ...             raise NotImplementedError()
    ...
    ...     @property
    ...     def key_size(self) -> int:
    ...         return self._cloud_client.key_size(self._key_id)
    ...
    ...     def public_key(self) -> rsa.RSAPublicKey:
    ...         raise NotImplementedError()
    ...
    ...     def private_numbers(self) -> rsa.RSAPrivateNumbers:
    ...         """
    ...         This method typically can't be implemented for cloud KMS keys
    ...         as the private key material is not accessible.
    ...         """
    ...         raise NotImplementedError()
    ...
    ...     def private_bytes(
    ...         self,
    ...         encoding: serialization.Encoding,
    ...         format: serialization.PrivateFormat,
    ...         encryption_algorithm: serialization.KeySerializationEncryption,
    ...     ) -> bytes:
    ...         """
    ...         This method typically can't be implemented for cloud KMS keys
    ...         as the private key material is not accessible.
    ...         """
    ...         raise NotImplementedError()
    ...
    ...     def __copy__(self) -> "CloudRSAPrivateKey":

>     ...         raise NotImplementedError()

Reviewer: Is there a reason this doesn't just `return self`

    ...
    >>> cloud_private_key = CloudRSAPrivateKey("creds", "key_id")
    >>> sig = cloud_private_key.sign(b"message", PKCS1v15(), hashes.SHA256())
    >>> isinstance(sig, bytes)
    True

> This key can then be used with other parts of the library, such as the X.509 APIs.

Reviewer: ```suggestion
This key can then be used with other parts of ``cryptography``, such as the X.509 APIs.
```
